contributingGuides/TYPESCRIPT_STYLE.md

```diff
@@ -0,0 +1,897 @@
+        // I ensured `e` is an `Error` object.
+        console.error(e.message);
+    }
+}
+```
+
+### Custom type guards
+
+A type guard is a check that narrows the type of a variable within a certain scope.
+
+The `typeof` keyword must be used when you need to narrow the type to primitive structures, like `string` or `number`.
+
+```ts
+function padLeft(value: string, padding: string | number) {
+    if (typeof padding === "number") {
+        return Array(padding + 1).join(" ") + value;
+    }
+    if (typeof padding === "string") {
+        return padding + value;
+    }
+    throw new Error(`Expected string or number, got '${padding}'.`);
+}
+
+padLeft("Hello world", 4); // returns "    Hello world"
+```
+
+We can also create user-defined type guards which are functions that perform a runtime check that guarantees the type in a certain scope. 
+
+```ts
+type PolicyRoomReport = {
+    type: 'policyRoom';
+    id: string;
+    roomId: string;
+};
+
+type PolicyExpenseChatReport = {
+    type: 'policyExpenseChat';
+    id: string;
+    expenseChatId: string;
+};
+
+type ChatReport = PolicyRoomReport | PolicyExpenseChatReport;
+
+function isPolicyRoomReport(report: ChatReport): report is PolicyRoomReport {
+    return report.type === 'policyRoom';
+}
+
+if (isPolicyRoomReport(report)) {
+    // I'm guaranteed to have access to PolicyRoomReport's properties e.g. "roomId".
+    // report.roomId
+} else {
+    // If this is not a PolicyRoomReport, then I'm guaranteed that this is a PolicyExpenseChatReport, so I'll have access to it's properties e.g. "expenseChatId"
+    // report.expenseChatId
+}
+```
+
+In this example, we have the `ChatReport` type which can be two possible types of object, `PolicyRoomReport` or `PolicyExpenseChatReport`. The `isPolicyRoomReport` function accepts `report` as a parameter and its return type is a [type predicate](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates), which means that it will narrow the type of `report` to `PolicyRoomReport` if the condition inside it is `true`, and to `PolicyExpenseChatReport` if the condition is `false`.
+
+## Typing arrays
+
+To type arrays, use `T[]` for simple types (i.e. types which are just primitive names or type references). Use `Array<T>` for all other types (union types, intersection types, object types, function types, etc).
+
+```ts
+// T[]
+const items: Item[] = [];
+const labels: string[] = ['name', 'city'];
+const indexes: number[] = [1, 2, 3];
+
+// Array<T>
+const items: Array<Partial<Item>> = [];
+const ids: Array<string | number> = ['some_id', 42];
+const colors: Array<ValueOf<typeof COLORS>> = [];
+```
+
+## Usage of “defaultProps”
+
+To achieve better and safer typing of components with default prop values, all usages `defaultProps` shall be removed and replaced by prop destructuring. Please head to **Typing components** section to understand how you can convert your implementation.
+
+## JSDoc annotations
```

Comment:
Can we also enforce this with a lint rule?

Reply:
I found this rule to disallow types on `@returns` and `@param` tags - [link](https://github.com/gajus/eslint-plugin-jsdoc/blob/main/docs/rules/no-types.md#readme). I did not find a rule to completely ban params and returns 😒

Reply:
@blazejkustra thank you for the research 🙌 

If I remember correctly, ESLint is stopping support for lining JSDoc 😨 . We can add it for now, but we should be careful not to rely on them overly.

Reply:
Done! cc @fabioh8010